### PR TITLE
`bpf_prog_attach` API and other changes.

### PR DESCRIPTION
**Pull request by shankarseal**
## Description

Depends on #871 

- Implements `bpf_prog_attach`
- Implements WFP callout for sock_addr hook.

## Testing

New test added that loads and attaches eBPF program of sock_addr type.

## Documentation

None.

### CODE REVIEW

File: include/ebpf_nethooks.h [hunk 1]
```diff
@@ -107,6 +107,10 @@ bind_hook_t(bind_md_t* context);
 //
 // CGROUP_SOCK_ADDR.
 //
+
+#define VERDICT_REJECT 0
+#define VERDICT_PROCEED 1
```
Comment (dthaler) [inline include/ebpf_nethooks.h:112]: Are these define names something that comes from Linux?   If so, fine.  If not, can we make them more unique, like BPF_SOCK_ADDR_VERDICT_REJECT or something similar?

File: tests/sample/cgroup_sock_addr.c [hunk 2]
```diff
@@ -44,7 +44,7 @@ authorize_v4(bpf_sock_addr_t* ctx)
 
     verdict = bpf_map_lookup_elem(&connection_policy_map, &tuple_key);
 
-    return (verdict != NULL) ? *verdict : 0;
+    return (verdict != NULL) ? *verdict : 1;
```
Comment (dthaler) [inline tests/sample/cgroup_sock_addr.c:47]:
```suggestion
    return (verdict != NULL) ? *verdict : VERDICT_PROCEED;
```

File: netebpfext/net_ebpf_ext_sock_addr.c [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+        classify_output->actionType = FWP_ACTION_PERMIT;
+        goto Exit;
+    }
+
+    _net_ebpf_extension_sock_addr_copy_wfp_connection_fields(incoming_fixed_values, &sock_addr_ctx);
+
+    if (net_ebpf_extension_hook_invoke_program(attached_client, &sock_addr_ctx, &result) != EBPF_SUCCESS)
+        goto Exit;
+
+    action = (result == VERDICT_PROCEED) ? FWP_ACTION_PERMIT : FWP_ACTION_BLOCK;
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext_sock_addr.c:407]: (just an observation that this treats any other value as block, so it need not be negative as the comments in the header said)

File: netebpfext/net_ebpf_ext_sock_addr.c [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+
+Exit:
+    if (attached_client)
+        net_ebpf_extension_hook_client_leave_rundown(attached_client, execution_type);
+
+    classify_output->actionType = action;
+
+    return;
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext_sock_addr.c:415]:
```suggestion
```

**Review by Alan-Jowett — APPROVED** [review]

File: include/ebpf_nethooks.h [hunk 5]
```diff
@@ -107,6 +107,10 @@ bind_hook_t(bind_md_t* context);
 //
 // CGROUP_SOCK_ADDR.
```
Comment (dthaler) [inline include/ebpf_nethooks.h:108]: Line 39 has xdp_hook_t.
Line 105 has bind_hook_t.
Where is the hook_t defined for this program type?

**Review by dthaler — DISMISSED** [review]

**Review by Alan-Jowett — DISMISSED** [review]

**Review by Alan-Jowett — APPROVED** [review]

**Review by dthaler — APPROVED** [review]

**Pull request merged**